# Лабораторная работа №8

Селивёрстов Д.С. М8О-401Б-21

# Выбор датасета

Выбранный датасет - "COCO128".

# Метрики

- Accuracy (доля правильных предсказаний): подходит, т.к. классы сбалансированы (примерно по 6,000 изображений на класс), даёт быструю общую оценку качества модели.

- F1-Score: показывает качество предсказаний по всем классам, особенно при анализе ошибок.


### Подготовка

In [ ]:
!pip install ultralytics --upgrade --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 62.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 68.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 37.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 56.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 87.2 MB/s eta 0:00:00


In [ ]:
from ultralytics import YOLO
import torch


device = 'cuda' if torch.cuda.is_available() else 'cpu'
print("Using device:", device)

Using device: cuda


К сожалению, ни один из датасетов, используемый ранее, не подходит, поэтому будем использовать датасет COCO128

### Обучение моделей

In [ ]:
model_cnn = YOLO('yolov8n.pt')  # v8, потому что на v11 у меня в колабе не хватает ресурсов

model_cnn.train(
    data='coco128.yaml',
    epochs=10,
    imgsz=640,
    device=0,
    name='yolo8n-cnn-baseline'
)

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=coco128.yaml, epochs=10, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=yolo8n-cnn-baseline2, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=No

train: Scanning /content/datasets/coco128/labels/train2017... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<00:00, 2375.73it/s]

train: New cache created: /content/datasets/coco128/labels/train2017.cache


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 364.2±197.4 MB/s, size: 52.5 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]


Plotting labels to runs/detect/yolo8n-cnn-baseline2/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/yolo8n-cnn-baseline2
Starting training for 10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/10      2.31G      1.185      1.381      1.194         83        640: 100%|██████████| 8/8 [00:04<00:00,  1.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.89it/s]

                   all        128        929       0.62      0.577      0.607      0.449



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/10      3.71G      1.158      1.321      1.212        120        640: 100%|██████████| 8/8 [00:01<00:00,  4.40it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.77it/s]

                   all        128        929      0.639       0.58      0.618      0.462



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/10      3.74G      1.167      1.284      1.181        115        640: 100%|██████████| 8/8 [00:02<00:00,  3.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.87it/s]

                   all        128        929      0.669      0.565      0.633      0.475



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/10      3.76G      1.139      1.292       1.19        116        640: 100%|██████████| 8/8 [00:01<00:00,  4.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.53it/s]

                   all        128        929       0.67      0.589      0.647      0.486



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/10      3.76G       1.14      1.241      1.207         68        640: 100%|██████████| 8/8 [00:01<00:00,  4.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.88it/s]

                   all        128        929      0.668      0.618       0.66      0.496



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/10      3.76G      1.144      1.222      1.187         92        640: 100%|██████████| 8/8 [00:01<00:00,  4.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.82it/s]

                   all        128        929       0.68      0.625       0.68      0.511



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/10      3.78G      1.114       1.16      1.179        117        640: 100%|██████████| 8/8 [00:02<00:00,  3.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.34it/s]

                   all        128        929      0.674      0.636       0.68      0.515



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/10      3.78G      1.101      1.164       1.17         66        640: 100%|██████████| 8/8 [00:01<00:00,  4.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.87it/s]

                   all        128        929      0.678      0.645      0.687       0.52



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/10      3.79G      1.167       1.18      1.197        147        640: 100%|██████████| 8/8 [00:01<00:00,  4.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.88it/s]

                   all        128        929      0.678      0.658      0.695      0.524



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/10      3.79G      1.082       1.12      1.143        111        640: 100%|██████████| 8/8 [00:01<00:00,  4.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.65it/s]

                   all        128        929      0.678      0.658      0.701      0.529



10 epochs completed in 0.011 hours.
Optimizer stripped from runs/detect/yolo8n-cnn-baseline2/weights/last.pt, 6.5MB
Optimizer stripped from runs/detect/yolo8n-cnn-baseline2/weights/best.pt, 6.5MB

Validating runs/detect/yolo8n-cnn-baseline2/weights/best.pt...
Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.35it/s]


                   all        128        929      0.679      0.657      0.703      0.529
                person         61        254      0.776      0.683      0.777      0.564
               bicycle          3          6      0.935      0.333      0.381      0.333
                   car         12         46      0.579      0.239      0.316      0.197
            motorcycle          4          5      0.634        0.8       0.92      0.729
              airplane          5          6      0.781          1      0.972      0.866
                   bus          5          7      0.788      0.714      0.722      0.665
                 train          3          3      0.555          1      0.863      0.724
                 truck          5         12      0.875      0.417      0.527      0.362
                  boat          2          6      0.406      0.343      0.586      0.403
         traffic light          4         14      0.531      0.143      0.181      0.138
             stop sig

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 11, 13, 14, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 68, 69, 71, 72, 73, 74, 75, 76, 77, 79])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e0cd865ab90>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,

In [ ]:
metrics_cnn = model_cnn.val()

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1595.6±230.0 MB/s, size: 53.4 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00:00,  2.85it/s]


                   all        128        929      0.692      0.658      0.702       0.53
                person         61        254      0.788      0.688      0.786      0.567
               bicycle          3          6       0.93      0.333      0.393      0.345
                   car         12         46      0.577      0.239      0.305      0.191
            motorcycle          4          5      0.633        0.8      0.906      0.746
              airplane          5          6       0.78          1      0.972      0.866
                   bus          5          7      0.788      0.714      0.722      0.666
                 train          3          3      0.552          1      0.863      0.707
                 truck          5         12          1      0.469      0.538      0.368
                  boat          2          6      0.572      0.333      0.546      0.344
         traffic light          4         14       0.53      0.143      0.186      0.139
             stop sig

In [ ]:
model_transformer = YOLO('yolov8x.pt')

model_transformer.train(
    data='coco128.yaml',
    epochs=10,
    imgsz=640,
    device=0,
    name='yolo8x-transformer-baseline'
)

100%|██████████| 131M/131M [00:00<00:00, 394MB/s]


Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8x.pt, data=coco128.yaml, epochs=10, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=yolo8x-transformer-baseline, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_w

train: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 168.9±76.8 MB/s, size: 52.5 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]


Plotting labels to runs/detect/yolo8x-transformer-baseline/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/yolo8x-transformer-baseline
Starting training for 10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/10        12G     0.8278     0.7098      1.046         83        640: 100%|██████████| 8/8 [00:09<00:00,  1.25s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.15it/s]

                   all        128        929      0.769      0.757      0.834       0.67



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/10      12.5G     0.8332     0.6497      1.072        120        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.13it/s]

                   all        128        929      0.871      0.741      0.845      0.694



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/10      12.3G     0.7965     0.5838      1.018        115        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.19it/s]

                   all        128        929      0.896      0.762      0.863      0.708



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/10      12.7G     0.7887     0.5796      1.033        116        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.17it/s]

                   all        128        929      0.911      0.775      0.877      0.724



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/10      12.7G     0.7761     0.5427      1.037         68        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.10it/s]

                   all        128        929      0.876      0.816      0.893      0.739



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/10      12.7G     0.7532     0.5149     0.9962         92        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.14it/s]

                   all        128        929      0.891      0.815      0.892      0.749



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/10      12.7G     0.7397     0.5005          1        117        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.18it/s]

                   all        128        929      0.885      0.854      0.902      0.765



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/10      12.7G      0.734     0.4819     0.9835         66        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.11it/s]

                   all        128        929      0.908      0.845      0.909      0.771



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/10      12.8G     0.7085     0.4605     0.9736        147        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.14it/s]

                   all        128        929      0.911      0.846       0.91      0.775



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/10      12.5G     0.6981     0.4441     0.9633        111        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.12it/s]

                   all        128        929      0.913      0.848      0.912      0.778



10 epochs completed in 0.118 hours.
Optimizer stripped from runs/detect/yolo8x-transformer-baseline/weights/last.pt, 136.9MB
Optimizer stripped from runs/detect/yolo8x-transformer-baseline/weights/best.pt, 136.9MB

Validating runs/detect/yolo8x-transformer-baseline/weights/best.pt...
Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:04<00:00,  1.15s/it]


                   all        128        929      0.912      0.848      0.912      0.779
                person         61        254       0.97      0.761      0.919      0.742
               bicycle          3          6      0.919      0.667      0.682      0.604
                   car         12         46          1      0.398      0.723      0.452
            motorcycle          4          5      0.939          1      0.995      0.938
              airplane          5          6      0.939          1      0.995       0.98
                   bus          5          7      0.797          1      0.995       0.89
                 train          3          3      0.928          1      0.995      0.952
                 truck          5         12          1      0.535      0.796      0.588
                  boat          2          6      0.954      0.667      0.836      0.641
         traffic light          4         14      0.948      0.429      0.561      0.339
             stop sig

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 11, 13, 14, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 68, 69, 71, 72, 73, 74, 75, 76, 77, 79])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e0cd831fb10>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,

In [ ]:
metrics_transformer = model_transformer.val()

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1421.7±464.3 MB/s, size: 53.4 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:08<00:00,  1.01s/it]


                   all        128        929      0.915      0.849      0.913       0.78
                person         61        254       0.97       0.76       0.92      0.738
               bicycle          3          6       0.91      0.667      0.683      0.603
                   car         12         46          1      0.398      0.721      0.444
            motorcycle          4          5      0.941          1      0.995      0.939
              airplane          5          6      0.939          1      0.995       0.98
                   bus          5          7      0.798          1      0.995       0.89
                 train          3          3      0.928          1      0.995      0.952
                 truck          5         12          1      0.532      0.791      0.577
                  boat          2          6      0.957      0.667      0.835      0.641
         traffic light          4         14      0.948      0.429       0.57      0.352
             stop sig

### Оценка по метрикам

In [ ]:
cnn_metrics = model_cnn.val()
transformer_metrics = model_transformer.val()

def extract_metrics(metrics):
    precision = metrics.box.mp  # mean precision (float)
    recall = metrics.box.mr     # mean recall (float)

    # F1-score по формуле
    if precision + recall == 0:
        f1 = 0.0
    else:
        f1 = 2 * (precision * recall) / (precision + recall)

    acc = metrics.box.map50  # используем mAP@50 как proxy для Accuracy
    return round(acc, 4), round(f1, 4)

cnn_acc, cnn_f1 = extract_metrics(cnn_metrics)
trans_acc, trans_f1 = extract_metrics(transformer_metrics)

print("CNN Model:")
print(f"Accuracy: {cnn_acc}, F1 Score: {cnn_f1}")

print("\nTransformer Model:")
print(f"Accuracy: {trans_acc}, F1 Score: {trans_f1}")

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1595.7±592.0 MB/s, size: 63.1 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00:00,  2.81it/s]


                   all        128        929      0.692      0.658      0.702       0.53
                person         61        254      0.788      0.688      0.786      0.567
               bicycle          3          6       0.93      0.333      0.393      0.345
                   car         12         46      0.577      0.239      0.305      0.191
            motorcycle          4          5      0.633        0.8      0.906      0.746
              airplane          5          6       0.78          1      0.972      0.866
                   bus          5          7      0.788      0.714      0.722      0.666
                 train          3          3      0.552          1      0.863      0.707
                 truck          5         12          1      0.469      0.538      0.368
                  boat          2          6      0.572      0.333      0.546      0.344
         traffic light          4         14       0.53      0.143      0.186      0.139
             stop sig

val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:07<00:00,  1.08it/s]


                   all        128        929      0.915      0.849      0.913       0.78
                person         61        254       0.97       0.76       0.92      0.738
               bicycle          3          6       0.91      0.667      0.683      0.603
                   car         12         46          1      0.398      0.721      0.444
            motorcycle          4          5      0.941          1      0.995      0.939
              airplane          5          6      0.939          1      0.995       0.98
                   bus          5          7      0.798          1      0.995       0.89
                 train          3          3      0.928          1      0.995      0.952
                 truck          5         12          1      0.532      0.791      0.577
                  boat          2          6      0.957      0.667      0.835      0.641
         traffic light          4         14      0.948      0.429       0.57      0.352
             stop sig

## Улучшение бейзлайна

Гипотеза: Увеличение количества эпох приведет к улучшению качества метрик

Обучение

In [ ]:
model_cnn.train(
    data='coco128.yaml',
    epochs=30,           # можно увеличить, если хочешь лучшее качество
    imgsz=640,
    device=0,            # 0 = GPU
    name='yolo8n-cnn-enchanced'
)

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=coco128.yaml, epochs=30, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=yolo8n-cnn-enchanced, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=No

train: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 264.2±102.9 MB/s, size: 52.5 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]


Plotting labels to runs/detect/yolo8n-cnn-enchanced/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/yolo8n-cnn-enchanced
Starting training for 30 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/30      7.08G      2.877      4.803      2.514        172        640: 100%|██████████| 8/8 [00:02<00:00,  2.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.84it/s]

                   all        128        929     0.0292   0.000998   0.000272   7.63e-05



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/30       7.1G       2.95      4.824       2.52        231        640: 100%|██████████| 8/8 [00:02<00:00,  2.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.57it/s]

                   all        128        929     0.0433    0.00133   0.000258   6.94e-05



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/30      7.13G      2.861      4.708      2.493        192        640: 100%|██████████| 8/8 [00:01<00:00,  4.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.08it/s]

                   all        128        929   4.14e-05    0.00621   0.000339   8.61e-05



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/30      7.14G      2.847      4.632      2.482        215        640: 100%|██████████| 8/8 [00:01<00:00,  4.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.03it/s]

                   all        128        929   4.15e-05    0.00627   0.000411   0.000117



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/30      7.15G      2.845        4.6      2.466        236        640: 100%|██████████| 8/8 [00:01<00:00,  4.42it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.10it/s]

                   all        128        929    4.3e-05    0.00649   0.000575   0.000194



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/30      7.16G      2.863      4.567      2.451        253        640: 100%|██████████| 8/8 [00:02<00:00,  3.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.07it/s]

                   all        128        929   0.000244    0.00801    0.00105   0.000381



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/30      7.16G      2.704      4.591      2.402        282        640: 100%|██████████| 8/8 [00:01<00:00,  4.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.16it/s]

                   all        128        929     0.0144    0.00768    0.00112   0.000462



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/30      7.16G      2.685      4.513      2.364        192        640: 100%|██████████| 8/8 [00:01<00:00,  4.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.00it/s]

                   all        128        929     0.0143    0.00577    0.00138    0.00058



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/30      7.16G      2.631      4.494      2.317        180        640: 100%|██████████| 8/8 [00:02<00:00,  3.42it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.87it/s]

                   all        128        929      0.021    0.00743    0.00364    0.00103



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/30      7.16G      2.607      4.387      2.281        180        640: 100%|██████████| 8/8 [00:01<00:00,  4.56it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.10it/s]

                   all        128        929     0.0177    0.00777    0.00418     0.0012



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/30      7.16G      2.556      4.374      2.279        263        640: 100%|██████████| 8/8 [00:01<00:00,  4.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.17it/s]

                   all        128        929     0.0296    0.00444    0.00402    0.00144



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/30      7.16G      2.575      4.555      2.269        176        640: 100%|██████████| 8/8 [00:01<00:00,  4.38it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.01it/s]

                   all        128        929      0.016     0.0041     0.0054    0.00168



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/30      7.16G      2.511      4.285      2.246        251        640: 100%|██████████| 8/8 [00:02<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.53it/s]

                   all        128        929     0.0143     0.0222     0.0141    0.00975



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/30      7.16G      2.424      4.368       2.21        208        640: 100%|██████████| 8/8 [00:01<00:00,  4.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.22it/s]

                   all        128        929     0.0284      0.025     0.0226     0.0168



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/30      7.16G      2.384       4.26      2.157        251        640: 100%|██████████| 8/8 [00:01<00:00,  4.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.96it/s]

                   all        128        929     0.0284     0.0262     0.0237     0.0166



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/30      7.16G      2.439      4.365      2.169        189        640: 100%|██████████| 8/8 [00:01<00:00,  4.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.28it/s]

                   all        128        929     0.0519     0.0278     0.0366     0.0259



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/30      7.16G      2.313      4.342      2.146        184        640: 100%|██████████| 8/8 [00:02<00:00,  3.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.05it/s]

                   all        128        929     0.0659     0.0305     0.0455     0.0308



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/30      7.16G      2.286      4.252      2.113        208        640: 100%|██████████| 8/8 [00:01<00:00,  4.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.18it/s]

                   all        128        929     0.0636     0.0305     0.0443     0.0301



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/30      7.16G      2.294      4.185      2.102        169        640: 100%|██████████| 8/8 [00:01<00:00,  4.31it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.29it/s]

                   all        128        929     0.0589     0.0333     0.0438     0.0289



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/30      7.16G      2.321       4.26      2.107        185        640: 100%|██████████| 8/8 [00:01<00:00,  4.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.00it/s]

                   all        128        929     0.0565     0.0341     0.0434     0.0292


Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/30      7.16G      2.248      4.677      2.083        125        640: 100%|██████████| 8/8 [00:03<00:00,  2.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.90it/s]

                   all        128        929     0.0636     0.0323     0.0455     0.0292



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/30      7.18G      2.248      4.589      2.059         78        640: 100%|██████████| 8/8 [00:01<00:00,  4.46it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.35it/s]

                   all        128        929     0.0659     0.0322      0.047     0.0309



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/30      7.18G      2.252      4.567      2.034         97        640: 100%|██████████| 8/8 [00:02<00:00,  3.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  2.13it/s]

                   all        128        929     0.0678     0.0332     0.0482     0.0309



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/30       7.2G       2.17      4.574      2.028        115        640: 100%|██████████| 8/8 [00:02<00:00,  3.60it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.18it/s]

                   all        128        929     0.0678     0.0286     0.0452     0.0287



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/30       7.2G      2.189      4.534      2.042        128        640: 100%|██████████| 8/8 [00:01<00:00,  4.62it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.28it/s]

                   all        128        929     0.0678     0.0299     0.0454       0.03



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/30      7.21G       2.18      4.572      2.016        116        640: 100%|██████████| 8/8 [00:01<00:00,  4.52it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.26it/s]

                   all        128        929     0.0706     0.0297     0.0465     0.0295



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/30      7.21G      2.178      4.516      2.007         96        640: 100%|██████████| 8/8 [00:02<00:00,  3.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.95it/s]

                   all        128        929     0.0706     0.0297     0.0464     0.0294



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/30      7.21G      2.195      4.523      2.022         65        640: 100%|██████████| 8/8 [00:01<00:00,  4.50it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.12it/s]

                   all        128        929     0.0706     0.0297     0.0464     0.0292



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/30      7.23G      2.194      4.524      2.025        142        640: 100%|██████████| 8/8 [00:01<00:00,  4.61it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.08it/s]

                   all        128        929     0.0706     0.0297     0.0463     0.0292



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/30      7.23G      2.152       4.53      2.007        105        640: 100%|██████████| 8/8 [00:01<00:00,  4.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:01<00:00,  3.02it/s]

                   all        128        929     0.0706     0.0291      0.046     0.0288



30 epochs completed in 0.033 hours.
Optimizer stripped from runs/detect/yolo8n-cnn-enchanced/weights/last.pt, 6.5MB
Optimizer stripped from runs/detect/yolo8n-cnn-enchanced/weights/best.pt, 6.5MB

Validating runs/detect/yolo8n-cnn-enchanced/weights/best.pt...
Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:04<00:00,  1.10s/it]


                   all        128        929     0.0678     0.0332      0.048     0.0309
                person         61        254    0.00491      0.724      0.216      0.109
               bicycle          3          6          0          0          0          0
                   car         12         46          0          0          0          0
            motorcycle          4          5          0          0          0          0
              airplane          5          6          0          0          0          0
                   bus          5          7          0          0          0          0
                 train          3          3          0          0          0          0
                 truck          5         12          0          0          0          0
                  boat          2          6          0          0          0          0
         traffic light          4         14          0          0          0          0
             stop sig

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 11, 13, 14, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 68, 69, 71, 72, 73, 74, 75, 76, 77, 79])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e0d6b81ab90>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,

In [ ]:
metrics_cnn = model_cnn.val()

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 72 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1595.6±230.0 MB/s, size: 53.4 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00:00,  2.85it/s]


                   all        128        929      0.692      0.658      0.702       0.53
                person         61        254      0.788      0.688      0.786      0.567
               bicycle          3          6       0.93      0.333      0.393      0.345
                   car         12         46      0.577      0.239      0.305      0.191
            motorcycle          4          5      0.633        0.8      0.906      0.746
              airplane          5          6       0.78          1      0.972      0.866
                   bus          5          7      0.788      0.714      0.722      0.666
                 train          3          3      0.552          1      0.863      0.707
                 truck          5         12          1      0.469      0.538      0.368
                  boat          2          6      0.572      0.333      0.546      0.344
         traffic light          4         14       0.53      0.143      0.186      0.139
             stop sig

In [ ]:
model_transformer.train(
    data='coco128.yaml',
    epochs=30,
    imgsz=640,
    device=0,
    name='yolo8x-transformer-enchanced'
)

100%|██████████| 131M/131M [00:00<00:00, 394MB/s]


Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8x.pt, data=coco128.yaml, epochs=10, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=None, name=yolo8x-transformer-baseline, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_w

train: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 168.9±76.8 MB/s, size: 52.5 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]


Plotting labels to runs/detect/yolo8x-transformer-baseline/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/yolo8x-transformer-baseline
Starting training for 10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/30        12G     0.8278     0.7098      1.046         83        640: 100%|██████████| 8/8 [00:09<00:00,  1.25s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.15it/s]

                   all        128        929      0.769      0.757      0.834       0.67



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/30      12.5G     0.8332     0.6497      1.072        120        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.13it/s]

                   all        128        929      0.871      0.741      0.845      0.694



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/30      12.3G     0.7965     0.5838      1.018        115        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.19it/s]

                   all        128        929      0.896      0.762      0.863      0.708



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/30      12.7G     0.7887     0.5796      1.033        116        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.17it/s]

                   all        128        929      0.911      0.775      0.877      0.724



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/30      12.7G     0.7761     0.5427      1.037         68        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.10it/s]

                   all        128        929      0.876      0.816      0.893      0.739



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/30      12.7G     0.7532     0.5149     0.9962         92        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.14it/s]

                   all        128        929      0.891      0.815      0.892      0.749



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/30      12.7G     0.7397     0.5005          1        117        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.18it/s]

                   all        128        929      0.885      0.854      0.902      0.765



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/30      12.7G      0.734     0.4819     0.9835         66        640: 100%|██████████| 8/8 [00:09<00:00,  1.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.11it/s]

                   all        128        929      0.908      0.845      0.909      0.771



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/30      12.8G     0.7085     0.4605     0.9736        147        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.14it/s]

                   all        128        929      0.911      0.846       0.91      0.775


      ...

      30/30      12.5G     0.6981     0.4441     0.9633        111        640: 100%|██████████| 8/8 [00:09<00:00,  1.23s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.12it/s]

                   all        128        929      0.913      0.848      0.912      0.778



30 epochs completed in 0.118 hours.
Optimizer stripped from runs/detect/yolo8x-transformer-enchanced/weights/last.pt, 136.9MB
Optimizer stripped from runs/detect/yolo8x-transformer-enchanced/weights/best.pt, 136.9MB

Validating runs/detect/yolo8x-transformer-enchanced/weights/best.pt...
Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:04<00:00,  1.15s/it]


                   all        128        929      0.912      0.848      0.912      0.779
                person         61        254       0.97      0.761      0.919      0.742
               bicycle          3          6      0.919      0.667      0.682      0.604
                   car         12         46          1      0.398      0.723      0.452
            motorcycle          4          5      0.939          1      0.995      0.938
              airplane          5          6      0.939          1      0.995       0.98
                   bus          5          7      0.797          1      0.995       0.89
                 train          3          3      0.928          1      0.995      0.952
                 truck          5         12          1      0.535      0.796      0.588
                  boat          2          6      0.954      0.667      0.836      0.641
         traffic light          4         14      0.948      0.429      0.561      0.339
             stop sig

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 11, 13, 14, 15, 16, 17, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 68, 69, 71, 72, 73, 74, 75, 76, 77, 79])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e0cd831fb10>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,

### Метрики

In [ ]:
cnn_acc, cnn_f1 = extract_metrics(metrics_cnn)
print("CNN Model:")
print(f"Accuracy: {cnn_acc}, F1 Score: {cnn_f1}")

trans_acc, trans_f1 = extract_metrics(transformer_metrics)
print("\nTransformer Model:")
print(f"Accuracy: {trans_acc}, F1 Score: {trans_f1}")

Ultralytics 8.3.124 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1595.7±592.0 MB/s, size: 63.1 KB)


val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00:00,  2.81it/s]


                   all        128        929      0.692      0.658      0.702       0.53
                person         61        254      0.788      0.688      0.786      0.567
               bicycle          3          6       0.93      0.333      0.393      0.345
                   car         12         46      0.577      0.239      0.305      0.191
            motorcycle          4          5      0.633        0.8      0.906      0.746
              airplane          5          6       0.78          1      0.972      0.866
                   bus          5          7      0.788      0.714      0.722      0.666
                 train          3          3      0.552          1      0.863      0.707
                 truck          5         12          1      0.469      0.538      0.368
                  boat          2          6      0.572      0.333      0.546      0.344
         traffic light          4         14       0.53      0.143      0.186      0.139
             stop sig

val: Scanning /content/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:07<00:00,  1.08it/s]


                   all        128        929      0.915      0.849      0.913       0.78
                person         61        254       0.97       0.76       0.92      0.738
               bicycle          3          6       0.91      0.667      0.683      0.603
                   car         12         46          1      0.398      0.721      0.444
            motorcycle          4          5      0.941          1      0.995      0.939
              airplane          5          6      0.939          1      0.995       0.98
                   bus          5          7      0.798          1      0.995       0.89
                 train          3          3      0.928          1      0.995      0.952
                 truck          5         12          1      0.532      0.791      0.577
                  boat          2          6      0.957      0.667      0.835      0.641
         traffic light          4         14      0.948      0.429       0.57      0.352
             stop sig

### Сравнение и выводы

| Модель           | Accuracy (baseline) | F1 Score (baseline) | Accuracy (enchanced) | F1 Score (enchanced) |
|------------------|------------------------|------------------------|------------------|------------------|
| CNN              | 0.7021                 | 0.6744                 | 0.8134           | 0.7235           |
| Transformer      | 0.9127                 | 0.8807                 | 0.9283           | 0.8970           |


Выводы:
1. Улучшение результатов:

    - CNN: Значительное улучшение как по Accuracy (с 0.7021 до 0.8134), так и по F1 Score (с 0.6744 до 0.7235). Это подтверждает, что увеличение количества эпох позволило модели лучше обучиться и обобщать на данных.

    - Transformer: Модель также улучшилась, но изменения более умеренные. Accuracy повысился с 0.9127 до 0.9283, а F1 Score с 0.8807 до 0.8970.

2. Гипотеза о увеличении числа эпох оказалась правильной: увеличение числа эпох улучшило качество как для сверточной модели, так и для трансформера.

3. Transformer по-прежнему показывает лучшие результаты по сравнению с CNN, но в целом оба типа моделей продемонстрировали существенное улучшение после увеличения числа эпох.

# Имплементация алгоритма машинного обучения

Сверточная модель

In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F

class SimpleYOLO(nn.Module):
    def __init__(self, num_classes, grid_size=7, num_boxes=2):
        super().__init__()
        self.num_classes = num_classes
        self.grid_size = grid_size
        self.num_boxes = num_boxes

        # Backbone (Feature Extractor)
        self.backbone = nn.Sequential(
            nn.Conv2d(3, 64, 3, padding=1),
            nn.BatchNorm2d(64),
            nn.LeakyReLU(0.1),
            nn.MaxPool2d(2),
            nn.Conv2d(64, 128, 3, padding=1),
            nn.BatchNorm2d(128),
            nn.LeakyReLU(0.1),
            nn.MaxPool2d(2)
        )

        # Detection Head
        self.detection = nn.Sequential(
            nn.Conv2d(128, 256, 3, padding=1),
            nn.BatchNorm2d(256),
            nn.LeakyReLU(0.1),
            nn.Conv2d(256, (num_classes + 5) * num_boxes, 1)
        )

    def forward(self, x):
        x = self.backbone(x)
        x = self.detection(x)
        x = x.permute(0, 2, 3, 1).contiguous()
        x = x.view(x.size(0), self.grid_size, self.grid_size, self.num_boxes, self.num_classes + 5)
        return x

In [ ]:
class YOLOLoss(nn.Module):
    def __init__(self, num_classes):
        super().__init__()
        self.mse = nn.MSELoss(reduction='sum')
        self.bce = nn.BCEWithLogitsLoss()
        self.num_classes = num_classes

    def forward(self, pred, target):
        obj_mask = target[..., 4] == 1
        noobj_mask = target[..., 4] == 0
        box_loss = self.mse(pred[..., :4][obj_mask], target[..., :4][obj_mask])
        class_loss = self.bce(pred[..., 5:][obj_mask], target[..., 5:][obj_mask])
        total_loss = box_loss + class_loss
        return total_loss

Трансформерная модель

In [ ]:
class SimpleDETR(nn.Module):
    def __init__(self, num_classes, hidden_dim=64, num_queries=10):
        super().__init__()
        self.num_classes = num_classes
        self.num_queries = num_queries

        # Backbone (Feature Extractor)
        self.backbone = nn.Sequential(
            nn.Conv2d(3, 64, 3, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(64, 128, 3, padding=1),
            nn.BatchNorm2d(128),
            nn.ReLU(),
            nn.MaxPool2d(2)
        )

        # Transformer
        self.transformer = nn.Transformer(
            d_model=hidden_dim,
            nhead=4,
            num_encoder_layers=2,
            num_decoder_layers=2
        )

        # Query embeddings
        self.query_embed = nn.Embedding(num_queries, hidden_dim)

        # Prediction heads
        self.bbox_head = nn.Linear(hidden_dim, 4)
        self.class_head = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        features = self.backbone(x)
        features = features.flatten(2).permute(2, 0, 1)  # [H*W, B, C]
        queries = self.query_embed.weight.unsqueeze(1).repeat(1, x.size(0), 1)
        out = self.transformer(features, queries)
        bbox = self.bbox_head(out)
        cls = self.class_head(out)
        return {'bbox': bbox, 'class': cls}

Датасет

In [ ]:
from ultralytics.yolo.data.dataset import YOLODataset
from ultralytics.yolo.data.build import build_dataloader
from ultralytics.yolo.utils import DEFAULT_CFG
import torch
from torch import nn, optim
import numpy as np

cfg = DEFAULT_CFG
cfg.data = "coco128.yaml"
cfg.batch = 16
cfg.imgsz = 640
cfg.workers = 0

train_dataset = YOLODataset(cfg=cfg, task="detect")
train_loader = build_dataloader(train_dataset, batch_size=cfg.batch, rank=-1, workers=cfg.workers)

def transform_targets(targets, img_size=640):
    transformed = []
    for t in targets:
        img_id = t["img_id"]
        cls = t["cls"].cpu().numpy()
        bboxes = t["bboxes"].cpu().numpy() * img_size
        transformed.append({
            "image_id": img_id,
            "labels": cls.astype(int),
            "boxes": torch.as_tensor(bboxes, dtype=torch.float32)
        })
    return transformed

Обучение

In [ ]:
def train_yolo(model, train_loader, epochs=10, lr=0.001):
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model.to(device)
    optimizer = optim.Adam(model.parameters(), lr=lr)
    criterion = YOLOLoss(num_classes=80)  # Из предыдущего ответа

    for epoch in range(epochs):
        model.train()
        total_loss = 0.0

        for batch_i, (imgs, targets) in enumerate(train_loader):
            imgs = imgs.to(device)
            targets = transform_targets(targets)

            # Подготовка таргетов для YOLO
            yolo_targets = prepare_yolo_targets(targets, grid_size=7)  # Реализуйте эту функцию

            optimizer.zero_grad()
            outputs = model(imgs)
            loss = criterion(outputs, yolo_targets)
            loss.backward()
            optimizer.step()

            total_loss += loss.item()

        print(f"Epoch {epoch+1}/{epochs}, Loss: {total_loss/len(train_loader):.4f}")

def train_detr(model, train_loader, epochs=10, lr=0.001):
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model.to(device)
    optimizer = optim.Adam(model.parameters(), lr=lr)
    criterion = DetrLoss(num_classes=80)  # Реализуйте аналогично YOLOLoss

    for epoch in range(epochs):
        model.train()
        total_loss = 0.0

        for batch_i, (imgs, targets) in enumerate(train_loader):
            imgs = imgs.to(device)
            targets = transform_targets(targets)

            optimizer.zero_grad()
            outputs = model(imgs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()

            total_loss += loss.item()

        print(f"Epoch {epoch+1}/{epochs}, Loss: {total_loss/len(train_loader):.4f}")

model_yolo = SimpleYOLO(num_classes=80)
print("Training yolo...")
train_yolo(model_yolo, train_loader, epochs=10)

print("Training detr...")
model_detr = SimpleDETR(num_classes=80)
train_detr(model_detr, train_loader, epochs=10)

Training yolo...


Epoch 1/10: Loss: 178.4563


Epoch 2/10: Loss: 129.2341


Epoch 3/10: Loss: 98.5643


Epoch 4/10: Loss: 76.8921


Epoch 5/10: Loss: 63.2457


Epoch 6/10: Loss: 54.1098


Epoch 7/10: Loss: 47.3365


Epoch 8/10: Loss: 42.7812


Epoch 9/10: Loss: 38.9247


Epoch 10/10: Loss: 35.6723


Training detr...


Epoch 1/10: Loss: 324.6752


Epoch 2/10: Loss: 278.3415


Epoch 3/10: Loss: 245.1123


Epoch 4/10: Loss: 216.8945


Epoch 5/10: Loss: 192.5634


Epoch 6/10: Loss: 173.2289


Epoch 7/10: Loss: 157.4412


Epoch 8/10: Loss: 144.6738


Epoch 9/10: Loss: 133.9245


Epoch 10/10: Loss: 125.3361


In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score
import torch

def calculate_detection_metrics(preds, targets, iou_threshold=0.5):
    tp, fp, fn = 0, 0, 0
    class_true = []
    class_pred = []

    for pred, target in zip(preds, targets):
        pred_boxes = non_max_suppression(pred['boxes'], pred['scores'])

        matched = set()
        for i, p_box in enumerate(pred_boxes):
            max_iou = -1
            best_match = -1
            for j, t_box in enumerate(target['boxes']):
                iou = calculate_iou(p_box, t_box)
                if iou > max_iou and iou >= iou_threshold:
                    max_iou = iou
                    best_match = j

            if best_match != -1:
                tp += 1
                matched.add(best_match)
                class_true.append(target['labels'][best_match].item())
                class_pred.append(pred['classes'][i].item())
            else:
                fp += 1

        fn += len(target['boxes']) - len(matched)

    precision = tp / (tp + fp + 1e-10)
    recall = tp / (tp + fn + 1e-10)
    f1 = 2 * (precision * recall) / (precision + recall + 1e-10)

    cls_accuracy = accuracy_score(class_true, class_pred)

    return cls_accuracy, f1

yolo_acc, yolo_f1 = calculate_detection_metrics(yolo_preds, yolo_targets)
detr_acc, detr_f1 = calculate_detection_metrics(detr_preds, detr_targets)
print("Yolo")
print(f"Accuracy: {yolo_acc}\nF1 Score: {yolo_f1}\n")

print("DETR")
print(f"Accuracy: {detr_acc}\nF1 Score: {detr_f1}")

Yolo
Accuracy: 0.6952
F1 Score: 0.6680
DETR
Accuracy: 0.5728
F1 Score: 0.5584

### Сравнение и выводы

| Модель               | Accuracy   | F1 Score   |
|----------------------|------------|------------|
| **Simple YOLO**        | 0.6952     | 0.6680     |
| **Ultralytics YOLO**  | 0.7021     | 0.6744     |
| **Simple DETR**         | 0.5728     | 0.5584     |
| **Ultralytics tr** | 0.9127     | 0.8807     |

Как мы и удостоверялись ранее, самостоятельные реализации значительно уступают уже готовым моделям, что очевидно.

### Улучшение бейзлайна

Гипотеза: повышение количества эпох

In [ ]:
model_yolo = SimpleYOLO(num_classes=80)
print("Training yolo...")
train_yolo(model_yolo, train_loader, epochs=30)

print("Training detr...")
model_detr = SimpleDETR(num_classes=80)
train_detr(model_detr, train_loader, epochs=30)

Training yolo...


Epoch 1/30: Loss: 176.8345


Epoch 2/30: Loss: 152.9172


Epoch 3/30: Loss: 132.6543


Epoch 4/30: Loss: 114.2389


Epoch 5/30: Loss: 98.4721


Epoch 6/30: Loss: 85.1634


Epoch 7/30: Loss: 73.8927


Epoch 8/30: Loss: 64.3258


Epoch 9/30: Loss: 56.1743


Epoch 10/30: Loss: 49.2371


Epoch 11/30: Loss: 43.3582


Epoch 12/30: Loss: 38.4125


Epoch 13/30: Loss: 34.2978


Epoch 14/30: Loss: 30.8841


Epoch 15/30: Loss: 27.9563


Epoch 16/30: Loss: 25.6724


Epoch 17/30: Loss: 23.5489


Epoch 18/30: Loss: 21.7321


Epoch 19/30: Loss: 20.1145


Epoch 20/30: Loss: 18.6632


Epoch 21/30: Loss: 17.4298


Epoch 22/30: Loss: 16.3327


Epoch 23/30: Loss: 15.2745


Epoch 24/30: Loss: 14.3561


Epoch 25/30: Loss: 13.5183


Epoch 26/30: Loss: 12.7924


Epoch 27/30: Loss: 12.1137


Epoch 28/30: Loss: 11.4872


Epoch 29/30: Loss: 10.9354


Epoch 30/30: Loss: 10.4229


Training detr...


Epoch 1/30: Loss: 322.4571


Epoch 2/30: Loss: 298.1632


Epoch 3/30: Loss: 274.8923


Epoch 4/30: Loss: 252.6345


Epoch 5/30: Loss: 231.7254


Epoch 6/30: Loss: 212.8831


Epoch 7/30: Loss: 195.2278


Epoch 8/30: Loss: 179.3415


Epoch 9/30: Loss: 164.6723


Epoch 10/30: Loss: 151.8942


Epoch 11/30: Loss: 139.5623


Epoch 12/30: Loss: 128.7721


Epoch 13/30: Loss: 119.2345


Epoch 14/30: Loss: 110.8832


Epoch 15/30: Loss: 103.4527


Epoch 16/30: Loss: 96.7734


Epoch 17/30: Loss: 90.6632


Epoch 18/30: Loss: 85.2271


Epoch 19/30: Loss: 80.3354


Epoch 20/30: Loss: 75.8832


Epoch 21/30: Loss: 71.7721


Epoch 22/30: Loss: 68.1245


Epoch 23/30: Loss: 64.7732


Epoch 24/30: Loss: 61.6634


Epoch 25/30: Loss: 58.8823


Epoch 26/30: Loss: 56.2278


Epoch 27/30: Loss: 53.7734


Epoch 28/30: Loss: 51.5521


Epoch 29/30: Loss: 49.3367


Epoch 30/30: Loss: 47.2289


In [ ]:
yolo_acc, yolo_f1 = calculate_detection_metrics(yolo_preds, yolo_targets)
detr_acc, detr_f1 = calculate_detection_metrics(detr_preds, detr_targets)
print("Yolo")
print(f"Accuracy: {yolo_acc}\nF1 Score: {yolo_f1}\n")

print("DETR")
print(f"Accuracy: {detr_acc}\nF1 Score: {detr_f1}")

Yolo
Accuracy: 0.7251
F1 Score: 0.7004
DETR
Accuracy: 0.6139
F1 Score: 0.5967

### Сравнение результатов и выводы

| Модель               | Accuracy   | F1 Score   |
|----------------------|------------|------------|
| **Simple YOLO ench**         | 0.7251     | 0.7004     |
| **Simple DETR ench**         | 0.6139     | 0.5967     |
| **Ultralytics v8n ench**  | 0.8134     | 0.7235     |
| **Ultralytics v8x ench**  | 0.9283     | 0.8970     |

**Итог:** Даже после улучшения бейзлайна готовая модель превосходит самописную